File: SuperGLU/Core/Config/ServiceLauncher.java
package Core.Config;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import Core.BaseMessagingNode;
import Core.MessagingGateway;
import Util.SerializationConvenience;
import Util.SerializationFormatEnum;

/**
 * This class is responsible for launching services based on a
 * ServiceConfiguration Object. It should also be used to shut down services
 * once that functionality is implemented.
 * 
 * In order for a service to use this, it will need to implement a constructor
 * that takes a ServiceConfiguration object.
 * 
 * @author auerbach
 *
 */

public class ServiceLauncher {

	private Map<String, BaseMessagingNode> services;

	public ServiceLauncher() {
		services = new HashMap<>();
	}

	public void launchAndConnectAllServices(ServiceConfigurations configs) {

		for (String key : configs.getServiceConfigurationMap().keySet()) {
			ServiceConfiguration config = configs.getServiceConfigurationMap().get(key);

			launchService(config);
		}

		for (String key : configs.getServiceConfigurationMap().keySet()) {
			ServiceConfiguration config = configs.getServiceConfigurationMap().get(key);
			connectService(config);
		}

	}

	public void launchService(ServiceConfiguration config) {
		try {
			Constructor<?> constructor = config.getType().getConstructor(ServiceConfiguration.class);
			BaseMessagingNode service = (BaseMessagingNode) constructor.newInstance(config);
			services.put(config.getId(), service);
		} catch (NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException
				| IllegalArgumentException | InvocationTargetException e) {
			e.printStackTrace();
		}

	}

	public void connectService(ServiceConfiguration config) {
		BaseMessagingNode service = services.get(config.getId());

		if (service != null) {

			for (String connectionId : config.getNodes()) {
				BaseMessagingNode connection = services.get(connectionId);

				if (connection != null)
					service.addNode(connection);
			}
		}
	}

	public ServiceConfigurations readConfigurationFromFile(String fileName) {
		String fileAsString = this.fileToString(fileName);

		ServiceConfigurations result = (ServiceConfigurations) SerializationConvenience.nativeizeObject(fileAsString,
				SerializationFormatEnum.JSON_FORMAT);

		return result;
	}

	private String fileToString(String fileName) {
		String result = null;

		try {
			FileReader fileIn = new FileReader(fileName);
			BufferedReader reader = new BufferedReader(fileIn);

			StringBuilder fileAsString = new StringBuilder();

			while (reader.ready()) {
				fileAsString.append(reader.readLine());
			}

			reader.close();

			result = fileAsString.toString();

		} catch (FileNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

		return result;
	}

	public Map<String, BaseMessagingNode> getServices() {
		return this.services;
	}

	public void stopService(String serviceName) {
		if (this.services.containsKey(serviceName)) {
			BaseMessagingNode serviceToStop = this.services.get(serviceName);

			for (BaseMessagingNode connection : serviceToStop.getNodes()) {
				// Remove connections to other services.
				serviceToStop.onUnbindToNode(connection);
				connection.onUnbindToNode(serviceToStop);
			}

			// If it's a gateway make sure to shut it down properly
			if (serviceToStop instanceof MessagingGateway) {
				MessagingGateway gateway = (MessagingGateway) serviceToStop;

				gateway.disconnect();
			}

			// finally remove it from the list of services
			this.services.remove(serviceName);
		}
	}

}

File: python_module/SuperGLU/Core/ServiceLauncher.py
'''
Created on Mar 12, 2018

 * This class is responsible for launching services based on a
 * ServiceConfiguration Object. It should also be used to shut down services
 * once that functionality is implemented.
 * <p>
 * In order for a service to use this, it will need to implement a constructor
 * that takes a ServiceConfiguration object.
 *

@author: auerbach
'''
import importlib
from SuperGLU.Util.Serialization import nativizeObject
from SuperGLU.Util import Serialization



class ServiceLauncher():
    
    services = {}
    
    def launchAndConnectAllServices(self, serviceConfigurations):
        for key in serviceConfigurations.getServiceConfigurationKeys():
            if key != "id":
                serviceConfiguration = serviceConfigurations.getServiceConfiguration(key)
                self.launchService(serviceConfiguration)
            
        for key in serviceConfigurations.getServiceConfigurationKeys():
            if key != "id":
                serviceConfiguration = serviceConfigurations.getServiceConfiguration(key)
                self.connectService(serviceConfiguration)
        
    
    def launchService(self, serviceConfiguration):
        classAndModule = serviceConfiguration.getType()
        splitName = classAndModule.split(sep=".", maxsplit=100)
        
        
        #small bit of annoyance:  the module name and class name are concatenated in the config file.
        #we'll have to split it up here
        ii=0
        moduleName=""
        while ii < len(splitName) - 1:
            moduleName = moduleName + splitName[ii]
            if ii < len(splitName) -2:
                moduleName = moduleName + "."
            ii = ii + 1
        
        className = splitName[len(splitName) - 1]
        module = importlib.import_module(moduleName)
        serviceClass = getattr(module, className)
        service = serviceClass(serviceConfiguration)
        self.services[serviceConfiguration.getId()] = service
        
        
    def connectService(self, serviceConfiguration):
        if serviceConfiguration.getId() in self.services:
            service = self.services[serviceConfiguration.getId()]
        
            if service is not None:
                for connectionId in serviceConfiguration.getNodes():
                    connection = self.services[connectionId]
                    if connection is not None:
                        service.addNodes([connection])
                    
                    
    def fileToString(self, fileName):
        result = ""
        file = open(fileName)
        
        if file is None:
            return None
        
        for line in file:
            result = result + line
        
        return result

    
    def readConfigurationFromFile(self, fileName, defaultFileName):
        fileAsString = self.fileToString(fileName)
        
        if fileAsString is None:
            fileAsString = self.fileToString(defaultFileName)
            
        
        result = nativizeObject(fileAsString, None, Serialization.JSON_STANDARD_FORMAT)
        return result
    
    def stopService(self, serviceName):
        if serviceName in self:
            serviceToStop = self[serviceName]
            connections = serviceToStop.getNodes()
            for connection in connections:
                serviceToStop.unregister(connection)
                connection.unregister(serviceToStop)
            
            del self.services[serviceName] 
    
    def stopAllServices(self):
        for serviceName in self.keys():
            self.stopService(serviceName)    